File: typescript/src/prepared.ts
// SPDX-License-Identifier: MIT
// Copyright (C) 2026 APlane Project LLC

import type {
  PreparedGroup,
  PreparedTransaction,
  SignRequest,
} from "./types.js";
import { PQ_SCHEME_FALCON1024, wireLogicSigResources } from "./types.js";
import { SignerError } from "./errors.js";
import { encodeLsigArgs, encodeTransaction, bytesToHex } from "./encoding.js";

function base64ToHex(value: string): string {
  if (typeof Buffer !== "undefined") {
    return Buffer.from(value, "base64").toString("hex");
  }
  const binary = atob(value);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytesToHex(bytes);
}

/**
 * Convert one prepared transaction slot to a signer SignRequest entry.
 */
export function preparedTransactionToSignRequest(
  prepared: PreparedTransaction,
): SignRequest {
  if (prepared.signedTransactionBase64) {
    const request: SignRequest = {
      signed_txn_hex: base64ToHex(prepared.signedTransactionBase64),
    };
    if (prepared.lsigResources) {
      request.lsig_resources = wireLogicSigResources(
        prepared.lsigResources,
        "prepared passthrough lsigResources",
      );
    }
    return request;
  }

  if (!prepared.transaction) {
    throw new SignerError("transaction is required");
  }

  const [txnBytesHex, txnSender] = encodeTransaction(prepared.transaction);
  if (!prepared.authAddress) {
    const request: SignRequest = { txn_bytes_hex: txnBytesHex };
    if (prepared.lsigResources) {
      request.lsig_resources = wireLogicSigResources(
        prepared.lsigResources,
        "prepared foreign lsigResources",
      );
    }
    if (prepared.pqScheme) {
      if (request.lsig_resources) {
        throw new SignerError("foreign transaction cannot specify both pqScheme and lsigResources");
      }
      if (prepared.pqScheme !== PQ_SCHEME_FALCON1024) {
        throw new SignerError(`unsupported pqScheme ${JSON.stringify(prepared.pqScheme)}`);
      }
      request.pq_scheme = prepared.pqScheme;
    }
    return request;
  }

  const request: SignRequest = {
    txn_bytes_hex: txnBytesHex,
    auth_address: prepared.authAddress,
    txn_sender: prepared.txnSender || txnSender,
  };
  if (prepared.lsigArgs) {
    request.lsig_args = encodeLsigArgs(prepared.lsigArgs);
  }
  if (prepared.appCallInfo) {
    request.app_call_info = prepared.appCallInfo;
  }
  return request;
}

/**
 * Convert a prepared group to signer SignRequest entries.
 */
export function preparedGroupToSignRequests(group: PreparedGroup): SignRequest[] {
  if (!group.transactions || group.transactions.length === 0) {
    throw new SignerError("prepared group is empty");
  }
  return group.transactions.map((item, index) => {
    try {
      return preparedTransactionToSignRequest(item);
    } catch (error) {
      if (error instanceof Error) {
        throw new SignerError(`prepared transaction ${index}: ${error.message}`);
      }
      throw error;
    }
  });
}

File: go/prepared.go
// SPDX-License-Identifier: MIT
// Copyright (C) 2026 APlane Project LLC

package aplane

import (
	"encoding/base64"
	"encoding/hex"
	"fmt"

	"github.com/algorand/go-algorand-sdk/v2/types"
)

// PreparedCheck records SDK-side preflight information collected while
// preparing a transaction intent.
type PreparedCheck struct {
	Name    string
	Status  string
	Message string
	Data    map[string]any
}

// PreparedTransaction is the SDK-side representation of one transaction slot
// after client preparation and before apsigner planning/signing.
type PreparedTransaction struct {
	Transaction             *types.Transaction
	AuthAddress             string
	TxnSender               string
	SignerKey               *KeyInfo
	LsigArgs                LsigArgs
	LsigSize                int
	AppCallInfo             *AppCallInfo
	SignedTransactionBase64 string
	Checks                  []PreparedCheck
}

// SignRequest converts the prepared slot to the signer wire request shape.
func (p PreparedTransaction) SignRequest() (SignRequest, error) {
	if p.SignedTransactionBase64 != "" {
		decoded, err := base64.StdEncoding.DecodeString(p.SignedTransactionBase64)
		if err != nil {
			return SignRequest{}, fmt.Errorf("invalid passthrough transaction: invalid base64: %w", err)
		}
		return SignRequest{SignedTxnHex: hex.EncodeToString(decoded)}, nil
	}

	if p.Transaction == nil {
		return SignRequest{}, fmt.Errorf("transaction is required")
	}

	txnBytesHex := hex.EncodeToString(encodeTxn(*p.Transaction))
	if p.AuthAddress == "" {
		req := SignRequest{TxnBytesHex: txnBytesHex}
		if p.LsigSize > 0 {
			req.LsigSize = p.LsigSize
		}
		return req, nil
	}

	txnSender := p.TxnSender
	if txnSender == "" {
		txnSender = p.Transaction.Sender.String()
	}
	req := SignRequest{
		AuthAddress: p.AuthAddress,
		TxnSender:   txnSender,
		TxnBytesHex: txnBytesHex,
		AppCallInfo: p.AppCallInfo,
	}
	if len(p.LsigArgs) > 0 {
		req.LsigArgs = make(map[string]string, len(p.LsigArgs))
		for name, value := range p.LsigArgs {
			req.LsigArgs[name] = hex.EncodeToString(value)
		}
	}
	return req, nil
}

// PreparedGroup is an ordered group of prepared transaction slots.
type PreparedGroup struct {
	Transactions []PreparedTransaction
	Checks       []PreparedCheck
}

// NewPreparedGroup returns a prepared group with the provided slots.
func NewPreparedGroup(transactions ...PreparedTransaction) PreparedGroup {
	return PreparedGroup{Transactions: transactions}
}

// SignRequests converts the group to signer request entries.
func (g PreparedGroup) SignRequests() ([]SignRequest, error) {
	if len(g.Transactions) == 0 {
		return nil, fmt.Errorf("prepared group is empty")
	}
	requests := make([]SignRequest, len(g.Transactions))
	for i, txn := range g.Transactions {
		req, err := txn.SignRequest()
		if err != nil {
			return nil, fmt.Errorf("prepared transaction %d: %w", i, err)
		}
		requests[i] = req
	}
	return requests, nil
}
